src/Servers/HttpSys/src/AsyncAcceptContext.cs

```diff
@@ -195,8 +177,8 @@ internal void AllocateNativeRequest(uint? size = null, ulong requestId = 0)
             //Debug.Assert(size != 0, "unexpected size");
 
             // We can't reuse overlapped objects
-            uint newSize = size.HasValue ? size.Value : DefaultBufferSize;
-            var backingBuffer = new byte[newSize + AlignmentPadding];
+            int newSize = checked((int)((size.HasValue ? size.Value : DefaultBufferSize) + AlignmentPadding));
+            var backingBuffer = ArrayPool<byte>.Shared.Rent(newSize);
```

Comment:
Addresses https://github.com/aspnet/AspNetCore/issues/5849

Reply:
The other common problem with these request buffers is that they can get pinned for a long time and cause memory fragmentation. Pool helps, but only to a point. Ideally we could use a large contiguous slab that was always pinned.

Reply:
Are these rented buffers guaranteed to be zero'd out?

Reply:
No, is it important? Can zero them out if needed

Reply:
Might be better to use the kestrel pool, like IIS does

Reply:
@benaadams how does it look if you measure this pooling change in isolation? I assume this is 90% of the impact from this PR?

The scope of this change makes it quite risky and we want to avoid destabilizing HttpSys when we're not doing a lot of active development and testing on it.

And yes, assume you need to zero out the buffer until proven otherwise.

Reply:
Zero'd it out.

>  how does it look if you measure this pooling change in isolation? I assume this is 90% of the impact from this PR?

Its not actually the pooling I care about though...

So the HostContext is a very chunky struct and gets passed around and stored in async statemachines; its already increased in size for 3.0 and improving the Hosting Logs https://github.com/aspnet/AspNetCore/pull/10102 would make that struct even bigger. So the idea is to additionally pool that on the "connection" and shrink it to a pointer size however as @davidfowl pointed out that would make all servers worse other than Kestrel https://github.com/aspnet/AspNetCore/pull/10180#issuecomment-491540286

The problem is the other servers have no concept of the "connection" so no where to store it; so if we change to pooling request objects they can then store both it and the HttpContext on the reusable requests; this was my main aim.

As it turns out the other servers are also a hotbed of allocations, so saving 48 bytes seems strange when HttpSys is also allocating 5KB+ a request hence the additional reduction.

Mainly I care about the HostContext though...

Reply:
> Its not actually the pooling I care about though...

I think we should also look at how this additional pooling affects memory usage when the server idles after load.

Making Hosting's Context struct smaller is nice, and we should look into making it smaller even if we don't end up pooling. It would likely create more GC pressure, but that might be an OK tradeoff.

Reply:
90/10. If we get 90% of the improvements in Http.Sys via the array pooling and only 10% from the HostContext, we'd rather take the pooling change here. It's simpler, lower risk, and far more impactful.

Reply:
> Making Hosting's Context struct smaller is nice, and we should look into making it smaller even if we don't end up pooling.

Then it moves to allocating.

It currently doesn't show up as an allocation as its stack copied and then "pool"ed in the `ProcessRequestsAsync` and `ProcessRequests` statemachines (i.e. its a field in the AsyncStateMachineBox so its allocation is amortized for the connection); however it pays a copy cost through each method that receives it due to its size and in the copying to the statemachines,

Moving to a pooled struct wrapped* object means its passed by register and has essentially no copy costs. (*struct wrapped to devirtualize the methods). However, the now heap context has to be stored somewhere to be reused; for Kestrel we can easily do that on the connection, but for other servers there isn't currently anywhere (which is also the same for `DefaultHttpContext` which is currently stored on the connection in Kestrel)

Reply:
> 90/10. If we get 90% of the improvements in Http.Sys via the array pooling and only 10% from the HostContext, we'd rather take the pooling change here.

Sounds like a trade ;)

Reply:
> Then it moves to allocating.

Correct. That's why I was saying it might be an OK tradeoff.

Reply:
Before this change, HttpSys currently boxes the struct anyway so.... ¯\\_(ツ)_/¯

Will do the HostContext change with Kestrel benefits, and the ArrayPool change for HttpSys as two PRs

Reply:
Ah, when you said you didn't care about the pooling, I thought you were saying the benefit of this was avoiding struct copies. I didn't realize HttpSys was already boxing anyway.

Reply:
> Ah, when you said you didn't care about the pooling

I might have been a bit strong :)

Boxing a single field + allocation of object *may* be similar to boxing large struct? Though haven't measured it.

Reply:
Context changes  #10319

Reply:
I'm not against pooling. It seems that even just pooling the request buffers would be a big win. I do want to understand which optimizations can be made independently of one another. If @Tratcher is good with fast tracking just request buffer pooling part of the change (with or without the kestrel pool) maybe we should just get that in now.